### PR TITLE
SR-IOV binding: Use network-info to map between the network and PCI address

### PR DESCRIPTION
**Pull request by ormergi**
### What this PR does
Make use of network-info annotation, introduced by https://github.com/kubevirt/kubevirt/pull/11618,  for SR-IOV VMs to map between SR-IOV network logical name and underlying device PCI address

## Before this PR:
The network-pci-map annotation used for mapping between SR-IOV network logical name and the underlying PCI address

## After this PR:
The network-info annotation is used for mapping between SR-IOV network logical name and the underlying PCI address.
The network-pci-map annotation will no longer be added to virt-launcher pods.

In a scenrio of Kubevirt upgrade, running VMs who's virt-launcher pods had network-pci-map annotation should not be disrupted.
As soon as such VM is migrated, the migration target pod will be created with network-info annotation and be passed down to virt-launcher up using the existing downward API mechanism.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
In a scenario network-info consist of multiple interfaces asserting the annotation value is flaky because the interfaces order may change by the marshaling.
Thus such tests were changed to assert unmarshaled network-info interfaces.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
The network-info annotation is now used for mapping between SR-IOV network and the underlying device PCI address
```

### CODE REVIEW

File: pkg/virt-controller/network/pod_annotations.go [hunk 1]
```diff
@@ -31,19 +31,35 @@ import (
...
-	newAnnotations := map[string]string{}
+	var err error
+
+	var sriovIfacePciMap map[string]*networkv1.DeviceInfo
 	if vmispec.SRIOVInterfaceExist(interfaces) {
-		networkPCIMapAnnotationValue := deviceinfo.CreateNetworkPCIAnnotationValue(
-			networks, interfaces, multusStatusAnnotation,
-		)
-		newAnnotations[deviceinfo.NetworkPCIMapAnnot] = networkPCIMapAnnotationValue
+		sriovIfacePciMap, err = deviceinfo.MapNetworkNameToDeviceInfo(networks, multusStatusAnnotation, interfaces)
```
Comment (EdDev) [inline pkg/virt-controller/network/pod_annotations.go:38]: I do not see the SR-IOV interfaces being filtered.
I think you need an equivalent `MapBindingPluginNetworkNameToDeviceInfo` helper, e.g. `MapSRIOVNetworkNameToDeviceInfo`.

File: pkg/virt-controller/network/pod_annotations.go [hunk 2]
```diff
@@ -31,19 +31,35 @@ import (
...
 		}
+	}
+
+	networkDeviceInfoMap := map[string]*networkv1.DeviceInfo{}
+	for k, v := range pluginIfacePciMap {
+		networkDeviceInfoMap[k] = v
+	}
+	for k, v := range sriovIfacePciMap {
+		networkDeviceInfoMap[k] = v
+	}
```
Comment (EdDev) [inline pkg/virt-controller/network/pod_annotations.go:58]: Please extract it to a helper that merges maps.

File: pkg/virt-controller/network/pod_annotations_test.go [hunk 3]
```diff
@@ -136,9 +152,16 @@ var _ = Describe("pod annotations", func() {
 				libvmi.InterfaceDeviceWithSRIOVBinding("doo"),
 			}
 			podAnnotationMap := network.GeneratePodAnnotations(networks, interfaces, networkStatus, bindingPlugins)
-			Expect(podAnnotationMap).To(Equal(map[string]string{
-				downwardapi.NetworkInfoAnnot:  `{"interfaces":[{"network":"foo","deviceInfo":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.2"}}}]}`,
-				deviceinfo.NetworkPCIMapAnnot: `{"doo":"0000:65:00.3"}`,
+			Expect(podAnnotationMap).To(HaveLen(1))
```
Comment (EdDev) [inline pkg/virt-controller/network/pod_annotations_test.go:155]: Why is this needed?

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go [hunk 4]
```diff
@@ -32,10 +34,17 @@ type PCIAddressWithNetworkStatusPool struct {
...
 	if err != nil {
-		return nil, fmt.Errorf("failed to unmarshal network-pci map %w", err)
+		return nil, fmt.Errorf("failed to unmarshal network-info annotation: %w", err)
+	}
+
+	networkPciMap := map[string]string{}
+	for _, iface := range networkInfo.Interfaces {
+		if iface.DeviceInfo != nil && iface.DeviceInfo.Pci != nil && iface.DeviceInfo.Pci.PciAddress != "" {
+			networkPciMap[iface.Network] = iface.DeviceInfo.Pci.PciAddress
+		}
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:47]: Why is this needed?
Is it because the SR-IOV interfaces have not been filtered at the controller?

**Review by EdDev — CHANGES_REQUESTED** [review]
Very nice.

Overall looks good.

Except for the inline comments, please go over the commit messages, I saw two typos, one at the 1st commit and one at the last one.

File: pkg/virt-controller/network/pod_annotations.go (hunk 1 above)
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations.go:38]: I followed the current state, note that SR-IOV interfaces were not filtered as well.
`sriovIfacePciMap` is bad name and doesnt reflect it.

If SR-IOV interfaces are filtered here, the end result is network-info annotation have only SR-IOV and binding-plugin with device-info interfaces.
Other interfaces that use different bindings will not be included.

File: pkg/virt-controller/network/pod_annotations_test.go (hunk 3 above)
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations_test.go:155]: We cannot use Equal because the annotation value consist of JSON array that is order could change, JSONMatch matcher doesnt help neither.

I added this len check to make sure no other annotation were added beside the network-info one this test  expects. I think it make it closer to the Equal matcher.

File: pkg/virt-controller/network/pod_annotations.go (hunk 2 above)
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations.go:58]: It seems that since Go 1.21 maps package is graduated and now part of the standard libs https://pkg.go.dev/maps@go1.21.0#Copy , we can use maps.Copy instead :slightly_smiling_face:

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go (hunk 4 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:47]: Previously the networkPciMap annotation was in form of map[string]string and it could be used as is.
The new network-info annotation is in different form and require some processing to get the relevant info (network to pci-address mapping)

File: pkg/virt-controller/network/pod_annotations.go (hunk 1 above)
Comment (EdDev) [reply pkg/virt-controller/network/pod_annotations.go]: > I followed the current state, note that SR-IOV interfaces were not filtered as well. `sriovIfacePciMap` is bad name and doesnt reflect it.

Per what I see, they have been filtered [1] and you removed it with this change.

[1] https://github.com/kubevirt/kubevirt/blob/a72353209753b177a2d71abc7204c6254befc89a/pkg/network/deviceinfo/sriov.go#L60

> If SR-IOV interfaces are filtered here, the end result is network-info annotation have only SR-IOV and binding-plugin with device-info interfaces. Other interfaces that use different bindings will not be included.

Who said they need to be included?
The PR is focused on SR-IOV, please keep it focused on it or reason for something else.
If you want to generalize it, it will be easier to open a different PR and convince us it makes sense.

File: pkg/virt-controller/network/pod_annotations.go (hunk 2 above)
Comment (EdDev) [reply pkg/virt-controller/network/pod_annotations.go]: Nice.
Still, please extract it to a helper.

**Comment by ormergi** [comment]
rebased

File: pkg/virt-controller/network/pod_annotations_test.go (hunk 3 above)
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations_test.go:155]: The HaveKey doesnt ensure there is one an only key, by asserting HaveLen(1) we make sure only the expected annotation exist.

Comment (EdDev) [reply pkg/virt-controller/network/pod_annotations_test.go:155]: Ok, but I do not like this one. I would have preferred a normalization to be done over this.
But it is not worth the effort.

File: pkg/network/deviceinfo/deviceinfo.go [hunk 5]
```diff
@@ -49,7 +49,7 @@ func MapBindingPluginNetworkNameToDeviceInfo(networks []v1.Network,
 	)
 }
 
-func mapNetworkNameToDeviceInfo(networks []v1.Network,
+func MapNetworkNameToDeviceInfo(networks []v1.Network,
```
Comment (AlonaKaplan) [inline pkg/network/deviceinfo/deviceinfo.go:52]: Please add unit tests to this method.

File: pkg/virt-controller/network/pod_annotations.go [hunk 6]
```diff
@@ -31,19 +33,31 @@ import (
...
-	newAnnotations := map[string]string{}
+	var err error
+
+	var ifacePciMap map[string]*networkv1.DeviceInfo
 	if vmispec.SRIOVInterfaceExist(interfaces) {
-		networkPCIMapAnnotationValue := deviceinfo.CreateNetworkPCIAnnotationValue(
-			networks, interfaces, multusStatusAnnotation,
-		)
-		newAnnotations[deviceinfo.NetworkPCIMapAnnot] = networkPCIMapAnnotationValue
+		ifacePciMap, err = deviceinfo.MapNetworkNameToDeviceInfo(networks, multusStatusAnnotation, interfaces)
```
Comment (AlonaKaplan) [inline pkg/virt-controller/network/pod_annotations.go:40]: Only the SRIOV interfaces should be added to the annotation.

File: pkg/virt-controller/network/pod_annotations.go [hunk 7]
```diff
@@ -31,19 +33,31 @@ import (
 )
 
 func GeneratePodAnnotations(networks []virtv1.Network, interfaces []virtv1.Interface, multusStatusAnnotation string, bindingPlugins map[string]virtv1.InterfaceBindingPlugin) map[string]string {
-	newAnnotations := map[string]string{}
+	var err error
+
+	var ifacePciMap map[string]*networkv1.DeviceInfo
 	if vmispec.SRIOVInterfaceExist(interfaces) {
-		networkPCIMapAnnotationValue := deviceinfo.CreateNetworkPCIAnnotationValue(
```
Comment (AlonaKaplan) [inline pkg/virt-controller/network/pod_annotations.go:36]: The old annotation should still be updated on existing virt-launcher pods. Otherwise, they will be broken.
SRIOV hotplug and maybe even migration will fail.

File: pkg/virt-controller/watch/vmi_test.go [hunk 8]
```diff
@@ -847,7 +845,16 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 			addActivePods(vmi, pod.UID, "")
 
 			controller.Execute()
-			expectPodAnnotations(pod, HaveKeyWithValue(deviceinfo.NetworkPCIMapAnnot, `{"`+sriovNetworkName+`":"`+selectedPCIAddress+`"}`))
+
+			Expect(pod.Annotations).To(HaveKey(downwardapi.NetworkInfoAnnot))
```
Comment (AlonaKaplan) [inline pkg/virt-controller/watch/vmi_test.go:849]: I would just check for the existence of the key, the rest is covered in the pod_annotations_test.

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go [hunk 9]
```diff
@@ -32,10 +34,17 @@ type PCIAddressWithNetworkStatusPool struct {
 func NewPCIAddressPoolWithNetworkStatus(networkPCIMapBytes []byte) (*PCIAddressWithNetworkStatusPool, error) {
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:34]: s/networkPCIMapBytes/networkInfoBytes

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go (hunk 4 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:47]: The network binding interfaces should be treated in the relevant sidecar. So I agree the SRIOV interface should be filtered. But I don't see a reason to generalize the package.

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go [hunk 10]
```diff
@@ -50,7 +50,7 @@ func CreateHostDevices(vmi *v1.VirtualMachineInstance) ([]api.HostDevice, error)
 	if len(SRIOVInterfaces) == 0 {
 		return []api.HostDevice{}, nil
 	}
-	netStatusPath := path.Join(downwardapi.MountPath, deviceinfo.NetworkPCIMapVolumePath)
+	netStatusPath := path.Join(downwardapi.MountPath, downwardapi.NetworkInfoVolumePath)
 	pciAddressPoolWithNetworkStatus, err := newPCIAddressPoolWithNetworkStatusFromFile(netStatusPath)
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:54]: That's the closest place to the real place I wanted to comment - 
`newPCIAddressPoolWithNetworkStatusFromFile` should be changed - s/networkPCIMapBytes/networkDeviceInfoBytes

**Review by ormergi — COMMENTED** [review]
@EdDev @AlonaKaplan thanks for review please take another look.

File: pkg/virt-controller/network/pod_annotations.go [hunk 11]
```diff
@@ -31,22 +33,43 @@ import (
...
 		newAnnotations[downwardapi.NetworkInfoAnnot] = networkDeviceInfoAnnotation
 	}
 
 	return newAnnotations
 }
+
+// mergeNetDeviceInfoMaps create unified map from the given map slice using maps.Copy.
+// When a key of a given map already exist in the unified one, the value in unified map
+// will be overwritten by the value associated with the key in the given map.
+func mergeMaps[M ~map[K]V, K comparable, V any](src ...M) M {
```
Comment (AlonaKaplan) [inline pkg/virt-controller/network/pod_annotations.go:69]: This generic form is hard to read and not required here. We know exactly what maps we are merging. It is an internal method that is not going to be used anywhere else.

File: pkg/network/deviceinfo/deviceinfo.go (hunk 5 above)
Comment (ormergi) [reply pkg/network/deviceinfo/deviceinfo.go:52]: Done

The code is changed to filter device-info plugin and SR-IOV interfaces and use `MapNetworkNameToDeviceInfo`

File: pkg/virt-controller/network/pod_annotations.go [hunk 12]
```diff
@@ -31,22 +33,43 @@ import (
...
-	newAnnotations := map[string]string{}
+	var err error
+
+	var sriovIfacePciMap map[string]*networkv1.DeviceInfo
 	if vmispec.SRIOVInterfaceExist(interfaces) {
-		networkPCIMapAnnotationValue := deviceinfo.CreateNetworkPCIAnnotationValue(
-			networks, interfaces, multusStatusAnnotation,
-		)
-		newAnnotations[deviceinfo.NetworkPCIMapAnnot] = networkPCIMapAnnotationValue
+		sriovIfaces := vmispec.FilterSRIOVInterfaces(interfaces)
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations.go:40]: Done

The code is changed to filter device-info plugin and SR-IOV interfaces and use `MapNetworkNameToDeviceInfo`

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go (hunk 10 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:54]: Sorry, I thought you mean for `NewPCIAddressPoolWithNetworkStatus` argument name.

Done.

File: pkg/virt-controller/network/pod_annotations.go (hunk 11 above)
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations.go:69]: I removed the merge-maps because its not longer necessary.

**Review by ormergi — COMMENTED** [review]
@AlonaKaplan @EdDev I resolved the last comments, also simplify the network device-info mapping code (now merging maps is not needed)
Please take another look

File: pkg/network/vmispec/interface_test.go [hunk 13]
```diff
@@ -238,7 +238,18 @@ var _ = Describe("VMI network spec", func() {
 			deviceInfoPlugin:    {DownwardAPI: v1.DeviceInfo},
 			nonDeviceInfoPlugin: {},
 		}
-
+		It("returns false when interface binding is not plugin with device-info", func() {
```
Comment (AlonaKaplan) [inline pkg/network/vmispec/interface_test.go:241]: The context is related to method `BindingPluginNetworkWithDeviceInfoExist`. The two added tests should have a separate context.

File: pkg/virt-controller/network/pod_annotations_test.go [hunk 14]
```diff
@@ -123,7 +126,20 @@ var _ = Describe("pod annotations", func() {
...
+
+			interfaces := []virtv1.Interface{
+				libvmi.InterfaceDeviceWithSRIOVBinding("foo"),
+			}
+			podAnnotationMap := network.GeneratePodAnnotations(networks, interfaces, networkStatus, bindingPlugins)
+			Expect(podAnnotationMap).To(Equal(map[string]string{
+				downwardapi.NetworkInfoAnnot: `{"interfaces":[{"network":"foo","deviceInfo":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.2"}}}]}`,
+			}))
+		})
+		It("should have network-info entry when there is SR-IOV interface and binding plugin interface with device-info", func() {
```
Comment (AlonaKaplan) [inline pkg/virt-controller/network/pod_annotations_test.go:142]: Since the interface filtering was moved to the `GeneratePodAnnotations` method please add more networks - non binding plugin, binding plugin with no device info to show those are filtered out.

File: pkg/network/vmispec/interface_test.go (hunk 13 above)
Comment (ormergi) [reply pkg/network/vmispec/interface_test.go:241]: One function depend on the other and duplicating the context with its vars and consts seems redundent.
Can we have an exception for those tests?

Comment (ormergi) [reply pkg/network/vmispec/interface_test.go:241]: Done

The const and vars were moved outside the original context to enable reuse.

File: pkg/virt-controller/network/pod_annotations_test.go (hunk 14 above)
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations_test.go:142]: Done

File: pkg/network/vmispec/interface.go [hunk 15]
```diff
@@ -174,10 +174,17 @@ func FilterInterfacesByNetworks(interfaces []v1.Interface, networks []v1.Network
...
-				return true
-			}
+		if BindingPluginDeviceInfoIface(iface, bindingPlugins) {
+			return true
+		}
+	}
+	return false
+}
+
+func BindingPluginDeviceInfoIface(iface v1.Interface, bindingPlugins map[string]v1.InterfaceBindingPlugin) bool {
```
Comment (EdDev) [inline pkg/network/vmispec/interface.go:184]: I do not understand from the name that it returns a bool.
Perhaps `HasBindingPluginDeviceInfo`.

File: pkg/network/vmispec/interface.go [hunk 16]
```diff
@@ -174,10 +174,17 @@ func FilterInterfacesByNetworks(interfaces []v1.Interface, networks []v1.Network
...
+	if iface.Binding != nil {
+		if binding, exist := bindingPlugins[iface.Binding.Name]; exist && binding.DownwardAPI == v1.DeviceInfo {
+			return true
```
Comment (EdDev) [inline pkg/network/vmispec/interface.go:187]: No condition is needed:
```go
binding, exist := bindingPlugins[iface.Binding.Name]
return exist && binding.DownwardAPI == v1.DeviceInfo
```

File: pkg/virt-controller/network/pod_annotations.go [hunk 17]
```diff
@@ -31,19 +29,18 @@ import (
...
 func GeneratePodAnnotations(networks []virtv1.Network, interfaces []virtv1.Interface, multusStatusAnnotation string, bindingPlugins map[string]virtv1.InterfaceBindingPlugin) map[string]string {
-	newAnnotations := map[string]string{}
-	if vmispec.SRIOVInterfaceExist(interfaces) {
-		networkPCIMapAnnotationValue := deviceinfo.CreateNetworkPCIAnnotationValue(
-			networks, interfaces, multusStatusAnnotation,
-		)
-		newAnnotations[deviceinfo.NetworkPCIMapAnnot] = networkPCIMapAnnotationValue
+	ifaces := vmispec.FilterInterfacesSpec(interfaces, func(iface virtv1.Interface) bool {
+		return iface.SRIOV != nil ||
+			vmispec.BindingPluginDeviceInfoIface(iface, bindingPlugins)
```
Comment (EdDev) [inline pkg/virt-controller/network/pod_annotations.go:34]: nit: Why two lines?

**Comment by AlonaKaplan** [comment]
Thanks!

/approve

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go [hunk 18]
```diff
@@ -32,9 +32,17 @@ import (
...
+		"interfaces": [
+			{"network":"pod"},
+			{"network":"network1", "deviceInfo":{
+				"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:04:02.5"}}},
+			{"network":"network2", "deviceInfo":{
+				"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:04:02.7"}}},
+			{"network":"network3", "deviceInfo":{
+				"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:04:02.8"}}}
+		]
+	}`)
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go:45]: I think `if iface.DeviceInfo != nil && iface.DeviceInfo.Pci != nil && iface.DeviceInfo.Pci.PciAddress != ""` opens up more combinations which are not covered here.

**Review by EdDev — COMMENTED** [review]
First commit:
The commit message has a typo, I think you meant `In` and not `If`.

Second and third commit:
You need to reason for dropping the prev SR-IOV annotation in the controller and virt-launcher. Explaining why it is ok with existing running VMs and migrations.

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go (hunk 18 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go:45]: Added unit tests to cover the mentioned cases.

**Review by ormergi — COMMENTED** [review]
> First commit:
> The commit message has a typo, I think you meant In and not If.
> 
> Second and third commit:
> You need to reason for dropping the prev SR-IOV annotation in the controller and virt-launcher. Explaining why it is ok with existing running VMs and migrations.

Done. 

@EdDev thanks for the review, PTAL

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go [hunk 19]
```diff
@@ -32,9 +32,17 @@ import (
 var _ = Describe("SRIOV PCI address pool with network-pci-map", func() {
 	var emptyFileBytes []byte
 	emptyNetworkPCIMapBytes := []byte("{}")
-	networkPCIMapWithThreeNetworks := []byte(
-		`{"network1":"0000:04:02.5","network2":"0000:04:02.7","network3":"0000:04:02.8"}`)
-
+	networkPCIMapWithThreeNetworks := []byte(`{
+		"interfaces": [
+			{"network":"pod"},
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go:37]: I do not think you need this here. It is confusing because such data is not expected to be placed in the annotation in the first place.

At the same time, the code protects against it but it is already tested directly through a dedicated test table.

**Review by EdDev — CHANGES_REQUESTED** [review]

File: pkg/network/deviceinfo/deviceinfo_test.go [hunk 20]
```diff
@@ -138,9 +111,12 @@ var _ = Describe("DeviceInfo", func() {
 			newBindingPluginInterface("doo", deviceInfoPlugin),
 		}
 		expectedMap := map[string]*networkv1.DeviceInfo{
-			"foo": {Type: "pci", Version: "1.0.0", Pci: &networkv1.PciDevice{PciAddress: "0000:65:00.2"}}}
-		Expect(deviceinfo.MapBindingPluginNetworkNameToDeviceInfo(
-			networks, interfaces, networkStatusWithMixedNetworks, bindingPlugins,
+			"default": nil,
+			"boo":     nil,
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/deviceinfo/deviceinfo_test.go:115]: Fixed, I changed the `MapNetworkNameToDeviceInfo` to not add non device-info plugin ifaces

File: pkg/network/vmispec/interface_test.go [hunk 21]
```diff
@@ -228,17 +228,36 @@ var _ = Describe("VMI network spec", func() {
...
+	bindingPlugins := map[string]v1.InterfaceBindingPlugin{
+		deviceInfoPlugin:    {DownwardAPI: v1.DeviceInfo},
+		nonDeviceInfoPlugin: {},
+	}
+	Context("binding plugin network with device info", func() {
+		It("returns false given non binding-plugin interface", func() {
+			Expect(netvmispec.HasBindingPluginDeviceInfo(
+				libvmi.InterfaceDeviceWithBridgeBinding("net1"),
+				bindingPlugins,
+			)).To(BeFalse())
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/vmispec/interface_test.go:245]: Nope, no error whatsoever.

File: pkg/virt-controller/network/pod_annotations.go [hunk 22]
```diff
@@ -31,19 +29,17 @@ import (
...
+		log.Log.Warningf("failed to create network device-info-map: %v", err)
 	}
-	if vmispec.BindingPluginNetworkWithDeviceInfoExist(interfaces, bindingPlugins) {
-		networkDeviceInfoMap, err := deviceinfo.MapBindingPluginNetworkNameToDeviceInfo(networks, interfaces, multusStatusAnnotation, bindingPlugins)
-		if err != nil {
-			log.Log.Warningf("failed to create network-device-info-map: %v", err)
-			networkDeviceInfoMap = map[string]*networkv1.DeviceInfo{}
-		}
+
+	newAnnotations := map[string]string{}
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations.go:40]: Done

File: pkg/virt-controller/network/pod_annotations_test.go [hunk 23]
```diff
@@ -123,22 +132,48 @@ var _ = Describe("pod annotations", func() {
...
+			networks := []virtv1.Network{
+				*libvmi.MultusNetwork("foo", "default/sriov"),
+			}
+
+			interfaces := []virtv1.Interface{
+				libvmi.InterfaceDeviceWithSRIOVBinding("foo"),
+			}
+			podAnnotationMap := network.GeneratePodAnnotations(networks, interfaces, networkStatus, bindingPlugins)
+			Expect(podAnnotationMap).To(Equal(map[string]string{
+				downwardapi.NetworkInfoAnnot: `{"interfaces":[{"network":"foo","deviceInfo":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.2"}}}]}`,
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-controller/network/pod_annotations_test.go:145]: Fixed

File: pkg/virt-controller/watch/vmi_test.go [hunk 24]
```diff
@@ -782,7 +781,7 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 			selectedPCIAddress = "0000:04:02.5"
 		)
 
-		It("should not patch network-pci-map and network-info when no SR-IOV/binding plugin networks exist", func() {
+		It("should not patch and network-info annotation when no SR-IOV/binding plugin networks exist", func() {
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-controller/watch/vmi_test.go:784]: Fixed

File: pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go (hunk 19 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go:37]: I removed the network-pod entry.

**Review by ormergi — COMMENTED** [review]
@EdDev thanks for the review, PTAL - [changes](https://github.com/kubevirt/kubevirt/compare/7a29ff3bc5ef753132e210942dc2acfd080ed971..cfc1aa02aa3f6e5316c81a88be2cb2502cc14168)

File: pkg/network/vmispec/interface_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/vmispec/interface_test.go:245]: The point was that you can make it an one-liner.

File: pkg/network/vmispec/interface_test.go (hunk 13 above)
Comment (EdDev) [reply pkg/network/vmispec/interface_test.go:247]: ?

**Review by EdDev — APPROVED** [review]
Thank you!
